### PR TITLE
Skeleton package

### PR DESCRIPTION
**Pull request by minrk**
This is a not-yet-working package that fills out the basic organization
for a REST App, largely based on JupyterHub itself

- html page handlers in handlers.py
- api handlers in apihandlers.py
- jinja templates
- CORS / XSRF defaults
- hub authentication
- hub services env
- package.json for installing the client-side with npm
- travis setup

### CODE REVIEW

**Comment by rgbkrk** [comment]
Is there a reason you're including `pbr`? Did that come from a cookiecutter template?

**Comment by minrk** [comment]
I've been giving pbr a try on a few small packages, and it's been pretty nice, so I thought I would give it a go here. It appears to offer the same declarative setup benefits of flit (use requirements.txt, add metadata to setup.cfg), but without any of the downsides (all standard pip, setup.py commands work, etc.).

File: README.md [hunk 1]
```diff
@@ -1,2 +1,19 @@
...
+- publish and download are explicit actions (no sharing in-place)
+
+Explicitly not goals (out of scope):
+
+- In-place or real-time sharing (these are tasks for the single-user server)
+
+This repo is in early development (it doesn't work).
+
+See [specification.md][specification.md] for the planned spec.
+
```
Comment (willingc) [inline README.md:19]: After merging this PR, I will circle back and edit this.

File: hubshare/app.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        url_s = os.environ.get('JUPYTERHUB_SERVICE_URL')
+        if not url_s:
+            return 9090
+        url = urlparse(url_s)
+        return url.port
+
+    template_paths = List(
+        help="Paths to search for jinja templates.",
+    ).tag(config=True)
+
```
Comment (willingc) [inline hubshare/app.py:98]: Consider removing line. I like the clean config blocks for `port`, `ip` above. Seems helpful to continue that style.

File: hubshare/app.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        """Exclude date from default date format"""
+        return "%Y-%m-%d %H:%M:%S"
+
+    @default('log_format')
+    def _log_format_default(self):
+        """override default log format to include time"""
+        return "%(color)s[%(levelname)1.1s %(asctime)s.%(msecs).03d %(name)s %(module)s:%(lineno)d]%(end_color)s %(message)s"
+
+    def init_logging(self):
+        # This prevents double log messages because tornado use a root logger that
```
Comment (willingc) [inline hubshare/app.py:122]: Let's add a simple docstring `"""Initialize logging""" for this to indicate that the comments are for `self.log.propagate = False`.

File: hubshare/app.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        logger.propagate = True
+        logger.parent = self.log
+        logger.setLevel(self.log.level)
+
+    def init_db(self):
+        """Initialize the HubShare database"""
+        self.db = None
+
+    def init_hub_auth(self):
+        self.hub_auth = HubAuth()
```
Comment (willingc) [inline hubshare/app.py:146]: Might as well add the docstring to this for future use of autodoc. `"""Initialize hub authentication"""`

File: hubshare/app.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        """Initialize tornado config"""
+        jinja_options = dict(
+            autoescape=True,
+        )
+        jinja_env = Environment(
+            loader=FileSystemLoader(self.template_paths),
+            **jinja_options
+        )
+
+        # if running from git, disable caching of require.js
```
Comment (willingc) [inline hubshare/app.py:158]: What does `if running from git` mean?

File: hubshare/app.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            jinja2_env=jinja_env,
+            version_hash=version_hash,
+            xsrf_cookies=True,
+        )
+        # allow configured settings to have priority
+        settings.update(self.tornado_settings)
+        self.tornado_settings = settings
+    
+    def init_handlers(self):
+        """Load our tornado request handlers"""
```
Comment (willingc) [inline hubshare/app.py:187]: our -> hubshare's

File: hubshare/handlers.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+"""handlers for human-facing pages"""
+
+from tornado import gen, web
+
+from jupyterhub.handlers import BaseHandler as JupyterHubBaseHandler
+from jupyterhub.services.auth import HubAuthenticated
+from jupyterhub.utils import url_path_join
+
+class BaseHandler(HubAuthenticated, JupyterHubBaseHandler):
+
```
Comment (willingc) [inline hubshare/handlers.py:10]: """A hubshare base handler."""

Add simple class docstring.

File: hubshare/handlers.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            static_url=self.static_url,
+            version_hash=self.version_hash,
+        )
+    
+    def finish(self):
+        return super(JupyterHubBaseHandler, self).finish()
+
+
+class Template404(BaseHandler):
+    """Render our 404 template"""
```
Comment (willingc) [inline hubshare/handlers.py:42]: our -> hubshare's

File: hubshare/handlers.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    urls = ['.*']
+
+    def prepare(self):
+        raise web.HTTPError(404)
+
+
+class RootHandler(BaseHandler):
+    urls = ['/']
```
Comment (willingc) [inline hubshare/handlers.py:50]: """Handler for serving hubshare's human facing pages"""

File: package.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  "dependencies": {},
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/jupyterhub/hubshare.git"
+  },
+  "keywords": [
+    "jupyter"
```
Comment (willingc) [inline package.json:14]: perhaps add "jupyterhub" too

**Review by willingc — APPROVED** [review]
Beautiful! Nice work Min.

I've left only minor edits/comments. Feel free to incorporate if time permits; otherwise, I can do them post-merge.

**Comment by willingc** [comment]
I'll handle the little stuff. Merging 🎆

**Pull request merged**